src/Microsoft.AspNet.Server.Kestrel/Networking/UvTcpHandle.cs

```diff
@@ -89,5 +108,48 @@ public static IPEndPoint CreateIPEndpoint(ServerAddress address)
 
             return new IPEndPoint(ip, address.Port);
         }
+
+        private static void GetIPAddressAndPort(Libuv.sockaddr sockaddr, out IPAddress address, out int port)
+        {
+            // The bytes are represented in network byte order.
+            //
+            // Example 1: [2001:4898:e0:391:b9ef:1124:9d3e:a354]:39179
+            //
+            // 0000 0000 0b99 0017  => The third and fourth bytes 990B is the actual port
+            // 9103 e000 9848 0120  => IPv6 address is represented in the 128bit field1 and field2.
+            // 54a3 3e9d 2411 efb9     Read these two 64-bit long from right to left byte by byte.
+            // 0000 0000 0000 0000
+            //
+            // Example 2: 10.135.34.141:39178
+            //
+            // 0000 0000 0a99 0017  => The port representation are the same
+            // 0000 0000 0000 0000
+            // 8d22 870a ffff 0000  => IPv4 occupies the last 32 bit: 0A.87.22.8d is the actual address. 
+            // 0000 0000 0000 0000
+            //
+            // IPAddress understands both IPv4 and IPv6.
+
+            var bytes1 = BitConverter.GetBytes(sockaddr._field1);
+            var bytes2 = BitConverter.GetBytes(sockaddr._field2);
+
+            var bytes = new byte[16];
+            for (int i = 0; i < 8; ++i)
+            {
+                bytes[i] = bytes1[i];
+                bytes[i + 8] = bytes2[i];
+            }
+
+            address = new IPAddress(bytes);
```

Comment:
Right now I assume the IPAddress performs well. If we detect later on it's a bottleneck we can replace it with our own bit-wise operation.